File: better/Models/Message.swift
import Foundation

struct Message: Codable, Identifiable, Equatable, Hashable {
    var id: String
    var role: String  // "user" or "model"
    var content: String
    var createdAt: Date
    var selectedAt: Date?
    var parentId: String?  // For branching - references another Message's id

    // Media (Firebase Storage URL)
    var mediaURL: String?
    var mediaMimeType: String?

    // Token counts
    var inputTokens: Int?
    var outputTokens: Int?
    var cachedTokens: Int?

    // Video generation cost (estimated from duration)
    var videoCost: Double?

    // Thinking content (for thinking models)
    var thinkingContent: String?

    init(
        id: String = UUID().uuidString,
        role: String,
        content: String,
        parentId: String? = nil
    ) {
        self.id = id
        self.role = role
        self.content = content
        self.createdAt = Date()
        self.parentId = parentId
    }

    /// Whether this message is from the user
    var isUser: Bool { role == "user" }

    /// Whether this message has an image attachment
    var hasImage: Bool { mediaMimeType?.hasPrefix("image/") == true }

    /// Whether this message has a PDF attachment
    var isPDF: Bool { mediaMimeType == "application/pdf" }

    /// Whether this message has any media attachment
    var hasMedia: Bool { mediaURL != nil }
}

File: web/src/models/Message.ts
export type MessageRole = 'user' | 'model';

export interface Message {
  id: string;
  role: MessageRole;
  content: string;
  createdAt: Date;
  selectedAt: Date | null;
  parentId: string | null;
  mediaURL: string | null;
  mediaMimeType: string | null;
  inputTokens: number | null;
  outputTokens: number | null;
  cachedTokens: number | null;
  thinkingContent: string | null;
}

export type CreateMessageInput = Omit<Message, 'id' | 'createdAt'>;

export type MessageWithMedia = Message & {
  mediaURL: string;
  mediaMimeType: string;
};

export interface TokenCounts {
  inputTokens: number;
  outputTokens: number;
  cachedTokens: number;
}

const generateId = (): string => {
  if (typeof crypto !== 'undefined' && typeof crypto.randomUUID === 'function') {
    return crypto.randomUUID();
  }

  return `msg_${Date.now()}_${Math.random().toString(16).slice(2)}`;
};

const createMessage = (role: MessageRole, content: string, parentId?: string | null): Message => ({
  id: generateId(),
  role,
  content,
  createdAt: new Date(),
  selectedAt: null,
  parentId: parentId ?? null,
  mediaURL: null,
  mediaMimeType: null,
  inputTokens: null,
  outputTokens: null,
  cachedTokens: null,
  thinkingContent: null,
});

export const createUserMessage = (content: string, parentId?: string | null): Message =>
  createMessage('user', content, parentId);

export const createModelMessage = (content: string, parentId?: string | null): Message =>
  createMessage('model', content, parentId);

export const hasMedia = (message: Message): boolean => Boolean(message.mediaURL);

export const hasTokecounts = (message: Message): boolean =>
  message.inputTokens != null || message.outputTokens != null || message.cachedTokens != null;
